iroh/src/node.rs

```diff
@@ -954,6 +955,121 @@ impl<D: BaoStore> RpcHandler<D> {
+            if let Err(e) = self.doc_import_file0(msg, tx).await {
+                tx2.send_async(DocImportProgress::Abort(e.into()))
+                    .await
+                    .ok();
+            }
+        });
+        rx.into_stream().map(DocImportFileResponse)
+    }
+
+    async fn doc_import_file0(
+        self,
+        msg: DocImportFileRequest,
+        progress: flume::Sender<DocImportProgress>,
+    ) -> anyhow::Result<()> {
+        use iroh_bytes::store::ImportMode;
+        use std::collections::BTreeMap;
+
+        let progress = FlumeProgressSender::new(progress);
+        let names = Arc::new(Mutex::new(BTreeMap::new()));
+        // convert import progress to provide progress
+        let import_progress = progress.clone().with_filter_map(move |x| match x {
+            ImportProgress::Found { id, name } => {
+                names.lock().unwrap().insert(id, name);
+                None
+            }
+            ImportProgress::Size { id, size } => {
+                let name = names.lock().unwrap().remove(&id)?;
+                Some(DocImportProgress::Found { id, name, size })
+            }
+            ImportProgress::OutboardProgress { id, offset } => {
+                Some(DocImportProgress::Progress { id, offset })
+            }
+            ImportProgress::OutboardDone { hash, id } => {
+                Some(DocImportProgress::IngestDone { hash, id })
+            }
+            _ => None,
+        });
+        let DocImportFileRequest {
+            doc_id,
+            author_id,
+            key,
+            path: root,
+            // todo: figure out tags
+            // in_place,
+            // tag,
+        } = msg;
+        // Check that the path is absolute and exists.
+        anyhow::ensure!(root.is_absolute(), "path must be absolute");
+        anyhow::ensure!(
+            root.exists(),
+            "trying to add missing path: {}",
+            root.display()
+        );
+
+        // todo: do we assume copy?
+        // let import_mode = match in_place {
+        //     true => ImportMode::TryReference,
+        //     false => ImportMode::Copy,
+        // };
+        let import_mode = ImportMode::Copy;
+
+        let (temp_tag, size) = self
+            .inner
+            .db
+            .import_file(root, import_mode, BlobFormat::Raw, import_progress)
+            .await?;
+
+        let hash_and_format = temp_tag.inner();
+        let HashAndFormat { hash, format } = *hash_and_format;
+        // todo: figure out tags
+        // let tag = match tag {
+        //     SetTagOption::Named(tag) => {
+        //         self.inner
+        //             .db
+        //             .set_tag(tag.clone(), Some(*hash_and_format))
+        //             .await?;
+        //         tag
+        //     }
+        //     SetTagOption::Auto => self.inner.db.create_tag(*hash_and_format).await?,
+        // };
+        let tag = self.inner.db.create_tag(*hash_and_format).await?;
```

Comment:
Ah, I think this we should *not* do :) This creates a permanent tag, so the blob cannot be garbage collected unless the tag is deleted. Adding the hash to the document already protects the blob from being garbage collected (until it is deleted or overwritten). So this line should be removed and instead the temp tag simply be dropped after `doc_set_hash()` completed.

@rklaehn is dropping temp tags enough or does it has to be deleted somehow to free the blob from tag protection?

Reply:
thank you, this is exactly the advice i needed, because i really don't have a grasp on the consequences of tags right now! figured i'd do it wrong and have someone yell at me about it, lol.